File: last/last v2/04-java-threads/Example8.java
import java.util.Arrays;

// =================================================================
//
// File: Example8.java
// Author: Pedro Perez
// Description: This file implements the enumeration sort algorithm.
// 				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================
/*----------------------------------------------------------------

*

* Multiprocesadores: Threads en Java.

* Fecha: 21-Sep-2021

* Autor: A01209400 Royer Donnet Arenas Camacho
		 A01654856 Hugo David Franco Avila

*

*--------------------------------------------------------------*/

public class Example8 extends Thread {
	
	private static final int SIZE = 100_000;
	private int array[],threadId;
	private int B[],C[];

	public Example8(int array[],int B[],int threadId) {
		this.array = array;
		this.threadId=threadId;
		this.B=B;
		C=new int[array.length];
	}

	public int [] getResult(){
		return B;
	}


	public void run(){
		int i=threadId,j;

		while(i<array.length){
			for(j=0;j<array.length;j++){
				if(array[i]>array[j]||array[i]==array[j]&&j<i){
					B[i]+=1;
				}
			}
			i+=Utils.MAXTHREADS;
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		int array[],C[],B[], block;
		Example8 threads[];
		double ms;

		array = new int[SIZE];
		C = new int[array.length];
		B = new int[array.length];
		Utils.randomArray(array);
		Utils.displayArray("before", array);

		block = SIZE / Utils.MAXTHREADS;
		threads = new Example8[Utils.MAXTHREADS];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int j = 1; j <= Utils.N; j++) {
			for(int i=0;i<array.length;i++){
				B[i]=0;
			}
			for (int i = 0; i < threads.length; i++) {
				threads[i] = new Example8(Arrays.copyOf(array, array.length),B,i);
			}

			startTime = System.currentTimeMillis();
			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			/** -------- */
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
					//System.out.println("joined thread" +i);
					
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms +=  (stopTime - startTime);

		}
		for(int i=0;i<array.length;i++){
			C[B[i]]=array[i];
		}

		for(int i=0;i<array.length;i++){
			array[i]=C[i];
		}
		Utils.displayArray("after", array);
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}

File: last/last v2/02-intro-cpp/example8.cpp
// =================================================================
//
// File: example5.cpp
// Author: Pedro Perez
// Description: This file contains the code that implements the
//				enumeration sort algorithm. The time this implementation
//				takes ill be used as the basis to calculate the
//				improvement obtained with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

/*----------------------------------------------------------------

*

* Multiprocesadores: C++

* Fecha: 3-Sep-2021

* Autores: A01209400 Royer Donnet Arenas Camacho
		   A01654856 Hugo David Franco Ávila

*

*--------------------------------------------------------------*/

#include <iostream>
#include <iomanip>
#include <cstring>
#include "utils.h"

const int SIZE = 100000; //1e5

using namespace std;

// implement your class here
class EnumSort
{
private:
	int *array, size;

public:
	EnumSort(int *a, int s) : array(a), size(s) { }

	void sort() {
		int *b = new int[size];
		int *c = new int[size];
		for (int i = 0; i < size; i++) {
			for (int j = 0; j < size; j++) {
				if (array[i] > array[j] || array[i] == array[j] && j < i) {
					b[i] += 1;
				}
			}
		}
		for(int i = 0; i < size; i++){
			c[b[i]] = array[i];
		}

		for(int i = 0; i < size; i++){
			array[i] = c[i];
		}
	}
};

int main(int argc, char *argv[])
{
	int *a;
	double ms;

	a = new int[SIZE];
	random_array(a, SIZE);
	display_array("before", a);
	int *aux = new int[SIZE];

	cout << "Starting..." << endl;
	ms = 0;
	for (int i = 0; i < N; i++) {
		start_timer();
		
		memcpy(aux, a, sizeof(int) * SIZE);
		EnumSort en(aux, SIZE);
		
		// call your method here.
		en.sort();

		ms += stop_timer();
	}

	memcpy(a, aux, sizeof(int) * SIZE);
	display_array("after", a);
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;
	delete[] a;
	delete [] aux;
	return 0;
}
